File: backend/validation.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

type competitionRule struct {
	MinMembers int
	MaxMembers int
}

type memberInput struct {
	Role  string
	Name  string
	Email string
	Phone string
}

type RegistrationInput struct {
	TeamName    string
	Institution string
	Competition string

	CaptainName  string
	CaptainEmail string
	CaptainPhone string

	Member1Name  string
	Member1Email string
	Member1Phone string

	Member2Name  string
	Member2Email string
	Member2Phone string

	// POD
	Member3Name  string
	Member3Email string
	Member3Phone string

	Member4Name  string
	Member4Email string
	Member4Phone string
}

var validCompetitions = map[string]bool{
	"SCML":       true,
	"POD":        true,
	"BCC":        true,
	"Petrosmart": true,
	"PPC":        true,
}

var competitionRules = map[string]competitionRule{
	"SCML":       {MinMembers: 1, MaxMembers: 2},
	"POD":        {MinMembers: 2, MaxMembers: 4},
	"BCC":        {MinMembers: 0, MaxMembers: 2},
	"Petrosmart": {MinMembers: 2, MaxMembers: 2},
	"PPC":        {MinMembers: 1, MaxMembers: 2},
}

func validateInput(input RegistrationInput) error {
	// validasi team
	if strings.TrimSpace(input.TeamName) == "" {
		return fmt.Errorf("nama tim wajib diisi")
	}
	if strings.TrimSpace(input.Institution) == "" {
		return fmt.Errorf("institusi wajib diisi")
	}
	if !validCompetitions[input.Competition] {
		return fmt.Errorf("kompetisi tidak valid")
	}

	// validasi captain
	if err := validateMember("Captain", input.CaptainName, input.CaptainEmail, input.CaptainPhone); err != nil {
		return err
	}

	rule, ok := competitionRules[input.Competition]
	if !ok {
		return fmt.Errorf("aturan jumlah member untuk kompetisi %s tidak ditemukan", input.Competition)
	}

	members := []memberInput{
		{Role: "Member 1", Name: input.Member1Name, Email: input.Member1Email, Phone: input.Member1Phone},
		{Role: "Member 2", Name: input.Member2Name, Email: input.Member2Email, Phone: input.Member2Phone},
		{Role: "Member 3", Name: input.Member3Name, Email: input.Member3Email, Phone: input.Member3Phone},
		{Role: "Member 4", Name: input.Member4Name, Email: input.Member4Email, Phone: input.Member4Phone},
	}

	providedMembers := 0
	for _, member := range members {
		if !isMemberProvided(member.Name, member.Email, member.Phone) {
			continue
		}

		if err := validateMember(member.Role, member.Name, member.Email, member.Phone); err != nil {
			return err
		}

		providedMembers++
	}

	if providedMembers < rule.MinMembers {
		return fmt.Errorf("kompetisi %s butuh minimal %d member (di luar captain)", input.Competition, rule.MinMembers)
	}

	if providedMembers > rule.MaxMembers {
		return fmt.Errorf("kompetisi %s maksimal %d member (di luar captain)", input.Competition, rule.MaxMembers)
	}

	return nil
}

func isMemberProvided(name, email, phone string) bool {
	_ = email
	_ = phone
	return strings.TrimSpace(name) != ""
}

func validateMember(role, name, email, phone string) error {
	if strings.TrimSpace(name) == "" {
		return fmt.Errorf("nama %s wajib diisi", role)
	}
	if !isValidEmail(email) {
		return fmt.Errorf("email %s tidak valid", role)
	}
	if !isValidPhone(phone) {
		return fmt.Errorf("nomor HP %s tidak valid (hanya angka, 10-13 digit)", role)
	}
	return nil
}

func isValidEmail(email string) bool {
	re := regexp.MustCompile(`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`)
	return re.MatchString(email)
}

func isValidPhone(phone string) bool {
	re := regexp.MustCompile(`^[0-9]{10,13}$`)
	return re.MatchString(phone)
}

File: src/lib/validation.ts
export function validateEmail(email: string): boolean {
  const re = /^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$/;
  return re.test(email);
}

export function validatePhone(phone: string): boolean {
  // Remove all non-digit characters (spaces, dashes, plus signs, etc.)
  const cleanedPhone = phone.replace(/\D/g, "");
  // Accept 10-13 digits (covers various Indonesian phone formats)
  const re = /^[0-9]{10,13}$/;
  return re.test(cleanedPhone);
}

export interface FormData {
  team_name: string;
  institution: string;
  competition: string;
  captain_name: string;
  captain_email: string;
  captain_phone: string;
  member1_name: string;
  member1_email: string;
  member1_phone: string;
  member2_name: string;
  member2_email: string;
  member2_phone: string;
  member3_name: string;
  member3_email: string;
  member3_phone: string;
  member4_name: string;
  member4_email: string;
  member4_phone: string;
}

export function validateRegistrationForm(formData: FormData): string | null {
  // Validate team info
  if (!formData.team_name.trim()) {
    return "Team name is required";
  }

  if (!formData.institution.trim()) {
    return "Institution is required";
  }

  if (!formData.competition) {
    return "Competition is required";
  }

  // Validate captain
  if (!formData.captain_name.trim()) {
    return "Captain name is required";
  }

  if (!validateEmail(formData.captain_email)) {
    return "Captain email is invalid";
  }

  if (!validatePhone(formData.captain_phone)) {
    return "Captain phone number is invalid (accept formats like: 08123456789, 0812-3456-789, +62-812-345-6789)";
  }

  const memberRules: Record<
    string,
    { required: number[]; optional: number[] }
  > = {
    SCML: { required: [], optional: [1, 2] },
    PPC: { required: [1], optional: [2] },
    Petrosmart: { required: [1, 2], optional: [] },
    BCC: { required: [], optional: [1, 2] },
    POD: { required: [1, 2, 3], optional: [4] },
  };

  const rule = memberRules[formData.competition] ?? {
    required: [],
    optional: [],
  };

  for (const i of rule.required) {
    const name =
      formData[`member${i}_name` as keyof FormData]?.toString().trim() || "";
    const email =
      formData[`member${i}_email` as keyof FormData]?.toString().trim() || "";
    const phone =
      formData[`member${i}_phone` as keyof FormData]?.toString().trim() || "";

    if (!name) {
      return `Member ${i} name is required`;
    }

    if (!validateEmail(email)) {
      return `Member ${i} email is invalid`;
    }

    if (!validatePhone(phone)) {
      return `Member ${i} phone number is invalid (accept formats like: 08123456789, 0812-3456-789, +62-812-345-6789)`;
    }
  }

  for (const i of rule.optional) {
    const name =
      formData[`member${i}_name` as keyof FormData]?.toString().trim() || "";
    const email =
      formData[`member${i}_email` as keyof FormData]?.toString().trim() || "";
    const phone =
      formData[`member${i}_phone` as keyof FormData]?.toString().trim() || "";

    // Check if any field is filled
    const hasAnyData = name || email || phone;

    // If no data filled, skip this optional member
    if (!hasAnyData) {
      continue;
    }

    // If any data is filled, all fields must be complete and valid
    if (!name) {
      return `Member ${i} name is required`;
    }

    if (!validateEmail(email)) {
      return `Member ${i} email is invalid`;
    }

    if (!validatePhone(phone)) {
      return `Member ${i} phone number is invalid (accept formats like: 08123456789, 0812-3456-789, +62-812-345-6789)`;
    }
  }

  // Non-POD competitions cannot submit member 3 or 4.
  if (formData.competition !== "POD") {
    const hasMember3 =
      formData.member3_name.trim() ||
      formData.member3_email.trim() ||
      formData.member3_phone.trim();
    const hasMember4 =
      formData.member4_name.trim() ||
      formData.member4_email.trim() ||
      formData.member4_phone.trim();
    if (hasMember3 || hasMember4) {
      return `Competition ${formData.competition} only allows up to Member 2`;
    }
  }

  return null;
}
